lib/rules/sort-destructure-keys.js

```diff
@@ -170,19 +144,76 @@ module.exports = {
     const sorter = createSorter(caseSensitive);
 
     return {
-      ObjectPattern(node) {
+      ObjectPattern(objectPatternNode) {
+        const scope = context.getScope();
+
+        function isReferencedByOtherProperties(id) {
+          for (const variable of scope.variables) {
+            if (variable.name !== id.name) {
+              continue;
+            }
+
+            for (const reference of variable.references) {
+              if (reference.identifier === id) {
+                continue;
+              }
+
+              let current = reference.identifier;
+              while (current) {
+                if (current === objectPatternNode) {
+                  return true;
+                }
+                current = current.parent;
+              }
+            }
+            return false;
+          }
+        }
+
+        /**
+         * Returns whether or not a node is safe to be sorted.
+         * @param {import('estree').ObjectPattern['properties'][number]} node
+         */
+        function shouldCheck(node) {
+          if (node.type !== "Property") {
+            return true;
+          }
+
+          switch (node.value.type) {
+            case "ObjectPattern":
+              return node.value.properties.every(shouldCheck);
+            case "ArrayPattern":
+              // Fake the element as a property for simplicity
+              return node.value.elements.every(
+                (e) => !e || shouldCheck({ type: "Property", value: e })
+              );
+            case "AssignmentPattern":
+              if (node.value.left.type === "Identifier") {
+                return !isReferencedByOtherProperties(node.value.left);
+              }
+              break;
+            case "Identifier": {
+              return !isReferencedByOtherProperties(node.value);
+            }
+            default:
+              return true;
+          }
+
+          return true;
```

Comment:
Let's make the `switch` be responsible for the final value by making every `case` return:

```suggestion
              return true;
            case "Identifier": {
              return !isReferencedByOtherProperties(node.value);
            }
            default:
              return true;
          }
```